invoices/invoiceregistry.go

```diff
@@ -745,46 +756,35 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
 
 	// If it isn't recorded, cancel htlc.
 	if !ok {
-		return &HodlEvent{
-			CircuitKey:   circuitKey,
-			AcceptHeight: currentHeight,
-		}, nil
+		return hodlEvent, nil
 	}
 
 	// Determine accepted height of this htlc. If the htlc reached the
 	// invoice database (possibly in a previous call to the invoice
 	// registry), we'll take the original accepted height as it was recorded
 	// in the database.
-	acceptHeight := int32(invoiceHtlc.AcceptHeight)
+	hodlEvent.AcceptHeight = int32(invoiceHtlc.AcceptHeight)
 
 	switch invoiceHtlc.State {
 	case channeldb.HtlcStateCanceled:
-		return &HodlEvent{
-			CircuitKey:   circuitKey,
-			AcceptHeight: acceptHeight,
-		}, nil
+		return hodlEvent, nil
 
 	case channeldb.HtlcStateSettled:
 		// Also settle any previously accepted htlcs. The invoice state
 		// is leading. If an htlc is marked as settled, we should follow
 		// now and settle the htlc with our peer.
-		for key, htlc := range invoice.Htlcs {
+		for _, htlc := range invoice.Htlcs {
 			if htlc.State != channeldb.HtlcStateSettled {
 				continue
 			}
 
-			i.notifyHodlSubscribers(HodlEvent{
-				CircuitKey:   key,
-				Preimage:     &invoice.Terms.PaymentPreimage,
-				AcceptHeight: int32(htlc.AcceptHeight),
-			})
+			hodlEvent.AcceptHeight = int32(htlc.AcceptHeight)
+			hodlEvent.Preimage = &invoice.Terms.PaymentPreimage
+
+			i.notifyHodlSubscribers(*hodlEvent)
 		}
 
-		return &HodlEvent{
-			CircuitKey:   circuitKey,
-			Preimage:     &invoice.Terms.PaymentPreimage,
-			AcceptHeight: acceptHeight,
-		}, nil
+		return hodlEvent, nil
```

Comment:
Preimage is now only set inside the loop. I think it is okay, but the logic becomes more implicit.

I think it would be good to at least split this commit in two. One commit that introduces the 'default' hodlEvent and another one that adds the `ResolutionResult`.

We need to be careful in this area to not break anything unnecessarily. The 'override defaults' approach isn't as explicit as how it was before. Maybe it is safer to have some kind of helper function that still makes it very clear what the hodl event is that is returned.

Reply:
I think that going with a `HodlEvent` constructor will be a better way to go, as you say default + override can introduce subtle changes.